### PR TITLE
Issue290 Make the rectNxM restart files valid again

### PR DESCRIPTION
**Pull request by timspainNERSC**
# Make the rectNxM restart files valid again
## Fixes \#290

---
# Change Description

After changes to the grid and the swap of the array ordering, the `init_rectNxM.nc` are no longer valid. 

---
# Test Description

The fixed files can be verified by the model running the config_rectNxM.cfg configuration files successfully.

### CODE REVIEW

**Comment by timspainNERSC** [comment]
To be merged after #306

**Comment by einola** [comment]
@timspainNERSC : Can you fix the conflicts on this one and request a review?

File: run/init_topaz128x128.py [hunk 1]
```diff
@@ -102,6 +102,8 @@ def topaz4_interpolate(target_lon_deg, target_lat_deg, data, lat_array):
     dgs_comp = datagrp.createDimension("dgstress_comp", n_dgstress)
     n_coords_comp = datagrp.createDimension("ncoords", n_coords)
     
+    hfield_dims = ("y", "x")
```
Comment (a-smith-github) [inline run/init_topaz128x128.py:105]: ditto / ditto

**Review by a-smith-github — CHANGES_REQUESTED** [review]
Review unfinished

But please more comments, especially in tests (thinking ModelArray_test in particular!!!) and notes for reviewers. Can't review effectively without them. Time is valuable. 


Could some of those run scripts be more unified to reduce maintenance?

**Comment by timspainNERSC** [comment]
PR #380 reflects what this PR will look like after PR #373 is merged into develop.

File: core/src/ParaGridIO.cpp [hunk 2]
```diff
@@ -203,7 +207,13 @@ void ParaGridIO::dumpModelState(
         }
         dimMap[type] = ncDims;
     }
-    // Everything that has components needs that dimension, too
+    // Reverse the order of the dimensions to translate between column-major ModelArray and
+    // row-major netCDF
+    for (auto& [type, v] : dimMap) {
+        std::reverse(v.begin(), v.end());
```
[the remark this answers is not in the document]
Comment (timspainNERSC) [reply core/src/ParaGridIO.cpp]: Entirely correct to flag this; the dimMap array was being reversed twice.

File: core/src/ParaGridIO.cpp [hunk 3]
```diff
@@ -115,6 +116,7 @@ ModelState ParaGridIO::getModelState(const std::string& filePath)
         if (newType == ModelArray::Type::Z) {
             std::vector<size_t> startVector(ModelArray::nDimensions(newType), 0);
             std::vector<size_t> extentVector = ModelArray::dimensions(newType);
+            std::reverse(extentVector.begin(), extentVector.end());
```
[the remark this answers is not in the document]
Comment (timspainNERSC) [reply core/src/ParaGridIO.cpp:119]: Added an extra explanatory comment.

**Review by TomMelt — APPROVED** [review]
Looks good to me.

**Pull request merged**